src/main/scala/viper/gobra/frontend/Desugar.scala

```diff
@@ -812,13 +812,45 @@ object Desugar {
 
           case b: PBlock => unit(blockD(ctx)(b))
 
-          case l: PLabeledStmt =>
+          case l: PLabeledStmt => {
             val proxy = labelProxy(l.label)
-            for {
-              _ <- declare(proxy)
-              _ <- write(in.Label(proxy)(src))
-              s <- goS(l.stmt)
-            } yield s
+            l.stmt match {
+              case forstmt : PForStmt => {
+                unit(block( // is a block because 'pre' might define new variables
+                  for {
+                    _ <- declare(proxy)
+                    _ <- write(in.Label(proxy)(src))
+                    dPre <- maybeStmtD(ctx)(forstmt.pre)(src)
+                    (dCondPre, dCond) <- prelude(exprD(ctx)(forstmt.cond))
+                    (dInvPre, dInv) <- prelude(sequence(forstmt.spec.invariants map assertionD(ctx)))
+                    (dTerPre, dTer) <- prelude(option(forstmt.spec.terminationMeasure map terminationMeasureD(ctx)))
+                    labelName = nm.pushFor(forstmt, info)
+
+                    dBody = blockD(ctx)(forstmt.body)
+                    loopLabelProxy = in.LabelProxy(labelName)(src)
+                    loopLabel = in.Label(loopLabelProxy)(src)
+                    dPost <- maybeStmtD(ctx)(forstmt.post)(src)
+
+                    _ = nm.popFor(forstmt, info)
+                    wh = in.Seqn(
+                      Vector(dPre) ++ dCondPre ++ dInvPre ++ dTerPre ++ Vector(
+                        in.While(dCond, dInv, dTer, in.Seqn(
+                          Vector(dBody, loopLabel, dPost) ++ dCondPre ++ dInvPre ++ dTerPre
+                        )(src), Some(l.label.name))(src)
+                      )
+                    )(src)
+                  } yield wh
+                ))
+              }
```

Comment:
Isn't this pretty much the same code as the code for desugaring a loop. Why do we need to special case this?

Reply:
Maybe I could make the repetitive code a function and call it